Lib/configparser.py

```diff
@@ -966,9 +966,9 @@ def __setitem__(self, key, value):
 
         # XXX this is not atomic if read_dict fails at any point. Then again,
         # no update method in configparser is atomic in this implementation.
-        if key == self.default_section:
+        if key == self.default_section and self._defaults != value:
             self._defaults.clear()
-        elif key in self._sections:
+        elif key in self._sections and self._sections[key] != value:
```

Comment:
This will slow down the general case because `!=` checks whether all keys and values in a mapping are the same.

The problem outlined in BPO-32108 only happens if `value` is a SectionProxy. Only in this case clearing the underlying dictionary will cause the subsequent `read_dict()` to fail.

So, instead of those checks, on line 967 make the following check:
```
if key in self and self[key] is value:
    return
```

`self[key]` returns a SectionProxy for the given section. If the `value` *is* that same proxy, we don't need to do anything.

Reply:
@ambv  Thanks for the suggestion and the explanation!